### PR TITLE
chore(argo-cd): Remove deprecated features and redesign Ingress

### PR DESCRIPTION
**Pull request by pdrastil**
It's more than 1 year since these were removed / deprecated and it's probably safe to drop backward compatibility.

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#versioning)
* [x] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#documentation)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md).
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/stable/developer-guide/ci/)).

### CODE REVIEW

**Review by tico24 — APPROVED** [review]

File: charts/argo-cd/templates/argocd-application-controller/statefulset.yaml [hunk 1]
```diff
@@ -13,8 +13,6 @@ metadata:
     {{- include "argo-cd.labels" (dict "context" . "component" .Values.controller.name "name" .Values.controller.name) | nindent 4 }}
 spec:
   replicas: {{ .Values.controller.replicas }}
-  # TODO: Remove for breaking release as history limit cannot be patched
-  revisionHistoryLimit: 5
```
Comment (mkilchhofer) [inline charts/argo-cd/templates/argocd-application-controller/statefulset.yaml:17]: If we drop this (and I'd sponsor to do so), we need a major bump and also instructions on howto upgrade.

File: charts/argo-cd/Chart.yaml [hunk 2]
```diff
@@ -3,7 +3,7 @@ appVersion: v2.9.3
 kubeVersion: ">=1.23.0-0"
 description: A Helm chart for Argo CD, a declarative, GitOps continuous delivery tool for Kubernetes.
 name: argo-cd
-version: 5.52.0
+version: 5.53.0
```
Comment (mkilchhofer) [inline charts/argo-cd/Chart.yaml:6]: Lets go for a major bump and completely fulfill the Semver spec ;) we don't need another project doing Semver with a custom deprecation policy 🙈

File: charts/argo-cd/templates/argocd-application-controller/statefulset.yaml (hunk 1 above)
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-application-controller/statefulset.yaml:17]: It's mentioned in README that you need to use replace strategy (`helm upgrade --force` or `helm template | kubectl replace -f -`)

File: charts/argo-cd/Chart.yaml (hunk 2 above)
Comment (pdrastil) [reply charts/argo-cd/Chart.yaml]: Bumped

**Comment by pdrastil** [comment]
@mkilchhofer If we want to do major bump - do we want to go through arguments and name things consistently across components. `env`, `envFrom` vs `extraArgs`.

**Review by yu-croco — APPROVED** [review]
🚀

**Comment by pdrastil** [comment]
Also another thing that might go into major release is the changes I've mentioned about multiple ingress controller implementations on Slack. Could be good opportunity to clean this up.

**Review by mbevc1 — APPROVED** [review]

**Review by yu-croco — APPROVED** [review]

File: charts/argo-cd/templates/argocd-application-controller/statefulset.yaml (hunk 1 above)
Comment (mkilchhofer) [reply charts/argo-cd/templates/argocd-application-controller/statefulset.yaml:17]: Helm upgrade force does not work anymore with helmv3+:
- https://github.com/argoproj/argo-helm/issues/970

But ok, we can release that👍

**Review by mkilchhofer — APPROVED** [review]

File: charts/argo-cd/templates/argocd-application-controller/statefulset.yaml (hunk 1 above)
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-application-controller/statefulset.yaml:17]: @mkilchhofer I'm going to port also #2415 into this as change in Ingress interface will be breaking

**Review by mbevc1 — APPROVED** [review]

**Comment by pdrastil** [comment]
Added breaking changes to ingress rewrite. As it's harder to review:
- Removed catch-all rule from all ingress resources
- Implementation is aligned with Bitnami chart (no options to supply paths or hosts by default)
   - There can be only single path for Argo CD server backend
   - There can be only one host that is configured for SSO login
   - Additional rules or hosts can be provided via `extraHosts`, `extraPaths` or `extraRules` 
   - TLS is now boolean option that expects well-known secret (argocd-server-tls) or custom TLS via `extraTls`
 - AWS & GKE Application Load Balancer options were moved into ingress and each have dedicated implementation that is enabled with `controller: aws` or `controller: gke`
   - GKE must always use port numbers to get valid health checks
   - ALB must have extra service for gRPC

File: charts/argo-cd/values.yaml [hunk 3]
```diff
@@ -1932,29 +1898,47 @@ server:
...
 
-    # -- List of ingress hosts
-    ## Argo Ingress.
-    ## Hostnames must be provided if Ingress is enabled.
-    ## Secrets must be manually created in the namespace
-    hosts: []
-      # - argocd.example.com
+    # -- Argo CD server hostname
+    ## NOTE: Hostname must be provided if Ingress is enabled
+    hostname: argocd.server.local
```
Comment (mkilchhofer) [inline charts/argo-cd/values.yaml:1921]: example.com

File: charts/argo-cd/values.yaml [hunk 4]
```diff
@@ -1964,17 +1948,70 @@ server:
...
-      # - secretName: your-certificate-name
-      #   hosts:
-      #     - argocd.example.com
-
-    # -- Uses `server.service.servicePortHttps` instead `server.service.servicePortHttp`
-    https: false
+    # -- Additional ingress rules
+    # @default -- `[]` (See [values.yaml])
+    extraRules: []
+      # - host: example.server.local
```
Comment (mkilchhofer) [inline charts/argo-cd/values.yaml:1954]: example.com

File: charts/argo-cd/values.yaml [hunk 5]
```diff
@@ -1987,32 +2024,27 @@ server:
...
     pathType: Prefix
+
+    # -- Enable TLS configuration for the hostname defined at `server.ingressGrpc.hostname`
+    ## TLS certificate will be retrieved from a TLS secret with name: `<hostname>-tls`
+    tls: false
+
+    # -- The list of additional hostnames to be covered by ingress record
+    # @default -- `[]` (See [values.yaml])
+    extraHosts: []
+      # - name: grpc.argocd.server.local
```
Comment (mkilchhofer) [inline charts/argo-cd/values.yaml:2043]: example.com

File: charts/argo-cd/values.yaml [hunk 6]
```diff
@@ -2022,15 +2054,25 @@ server:
       #       port:
       #         name: use-annotation
 
-    # -- Ingress TLS configuration for dedicated [gRPC-ingress]
-    tls: []
+    # -- Additional ingress rules
+    # @default -- `[]` (See [values.yaml])
+    extraRules: []
+      # - host: example.server.local
```
Comment (mkilchhofer) [inline charts/argo-cd/values.yaml:2060]: example.com

File: charts/argo-cd/values.yaml [hunk 7]
```diff
@@ -2798,6 +2754,69 @@ applicationSet:
...
+    labels: {}
+    # -- Additional ingress annotations
+    annotations: {}
+
+    # -- Defines which ingress ApplicationSet controller will implement the resource
+    ingressClassName: ""
+
+    # -- Argo CD ApplicationSet hostname
+    ## NOTE: Hostname must be provided if Ingress is enabled
+    hostname: argocd.server.local
```
Comment (mkilchhofer) [inline charts/argo-cd/values.yaml:2772]: example.com

**Review by mkilchhofer — CHANGES_REQUESTED** [review]
Can we use the example.com domain?
-> argocd.example.com

example.com is there for documentation and examples, it is an official reserved domain by IANA:
- https://github.com/argoproj/argo-helm/pull/2352
- https://github.com/argoproj/argo-helm/pull/2350
- IANA RFC https://www.rfc-editor.org/rfc/rfc2606.html

File: charts/argo-cd/values.yaml [hunk 8]
```diff
@@ -1934,29 +1900,47 @@ server:
...
 
+  # Argo CD server ingress configuration
   ingress:
     # -- Enable an ingress resource for the Argo CD server
     enabled: false
-    # -- Additional ingress annotations
-    annotations: {}
+    # -- Specific implementation for ingress controller. One of `generic`, `aws` or `gke`
+    ## Additional configuration might be required in related configuration sections
+    controller: generic
```
Comment (pdrastil) [inline charts/argo-cd/values.yaml:1909]: @mkilchhofer Any opinion on this parameter name that enables specific implementation for backing ingress controller (GKE, AWS maybe something else in the future that have kinda unique setup)

Comment (mkilchhofer) [reply charts/argo-cd/values.yaml:1909]: Also gave some thoughts about this but didn't found a better name 😁 
names that came in to my mind:
- controller
- type
- flavour
- variant

But I'd say, lets go with `controller`. 👍

**Comment by pdrastil** [comment]
@mkilchhofer Are you OK with current state? Merge is blocked since your last review.

File: charts/argo-cd/templates/argocd-application-controller/statefulset.yaml (hunk 1 above)
Comment (pdrastil) [reply charts/argo-cd/templates/argocd-application-controller/statefulset.yaml:17]: Changed to configuration in values.yaml to allow upgrade without replacing the resource based on discussion with @mkilchhofer

File: charts/argo-cd/values.yaml [hunk 9]
```diff
@@ -577,6 +577,9 @@ controller:
   # Additional replicas will cause sharding of managed clusters across number of replicas.
   replicas: 1
 
+  # -- Maximum number of controller revisions that will be maintained in controller StatefulSet history
+  revisionHistoryLimit: 5
```
[the remark this answers is not in the document]
Comment (pdrastil) [reply charts/argo-cd/values.yaml:581]: Problem that I've mentioned to @mkilchhofer is if you change this to different value like null `~` you have to replace whole deployment as this field is immutable. That's why the value is original 5 for upgrade and you can unset this / or use global one.

**Review by jmeridth — APPROVED** [review]

**Review by mkilchhofer — APPROVED** [review]

**Comment by jetersen** [comment]
Release note did not describe to what degree was changed nor provide an upgrade guide 😓
I read the release notes thinking you only changed ingress but even that broke because of the use of `hostname` which seems unconventional. Very common to use `hosts`
Now left to figure out what changes I had not migrated over the years.

**Comment by jetersen** [comment]
Oh overriding the configs.cm: breaks this condition:

```
{{- if eq (toString (index (coalesce .Values.server.config .Values.configs.cm) "exec.enabled")) "true" }}
```

```
Failed to load target state: failed to generate manifest for source 1 of 1: rpc error: code = Unknown desc = Manifest generation error (cached): `helm template . --name-template argocd --namespace argocd --kube-version 1.28 <api versions removed> --include-crds` failed exit status 1: Error: template: argocd-proxy/charts/argo-cd/templates/argocd-server/role.yaml:45:66: executing "argocd-proxy/charts/argo-cd/templates/argocd-server/role.yaml" at <.Values.configs.cm>: nil pointer evaluating interface {}.cm Use --debug flag to render out invalid YAML
```

**Comment by pdrastil** [comment]
@jetersen Hi thanks for feedback. I'll add additional info to release notes. I agree that `hostname` is uncommon, however it makes sense as all components share single domain and some configuration options (SSO, notifications) allow only single URL for callbacks. There is upcoming PR https://github.com/argoproj/argo-helm/pull/2499 that simplifies configuration of all services with this cross-cutting concern from single place and user config will look more like this sample, instead of copy & pasting same value into multiple places.

```yaml
global:
  domain: argocd.example.com

server:
  certificate:
    enabled: true
  ingress:
    enabled: true
    tls: true
```

**Comment by pdrastil** [comment]
@jetersen Have you seen this or you checked release notes that are part of the chart? https://github.com/argoproj/argo-helm/tree/main/charts/argo-cd#600

**Comment by jetersen** [comment]
Usually Renovate will pull the changelogs from ArtifactHub but because your pushing GitHub Releases as well Renovate, we basically do not see the changelog. So me just trusting GitHub Releases missed that you maintain the changelog in README.md
I wonder why the GitHub Releases do not contain the content from the changelog.md
Shouldn't just be a part of the github workflow for releasing? 🤔

**Pull request merged**